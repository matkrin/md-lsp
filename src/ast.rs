use markdown::{
    mdast::{FootnoteReference, Heading, Link, LinkReference, Node, Text},
    unist::{Point, Position},
};
use regex::Regex;

struct Extracted {
    content: String,
    start_position: usize,
    line_number: usize,
}

impl Extracted {
    fn link_text_node(&self, start_line: usize) -> Node {
        let link_text = Text {
            value: self.content.clone(),
            position: Some(Position {
                start: Point {
                    line: start_line + self.line_number,
                    column: self.start_position,
                    offset: self.start_position,
                },
                end: Point {
                    line: start_line + self.line_number,
                    column: self.start_position + self.content.len(),
                    offset: self.start_position + self.content.len(),
                },
            }),
        };
        Node::Text(link_text)
    }

    fn link_node(&self, start_line: usize) -> Node {
        let link_text_node = self.link_text_node(start_line);

        let link = Link {
            children: vec![link_text_node],
            position: Some(Position {
                start: Point {
                    line: start_line + self.line_number,
                    column: self.start_position - 2,
                    offset: self.start_position - 2,
                },
                end: Point {
                    line: start_line + self.line_number,
                    column: self.start_position + self.content.len() + 2,
                    offset: self.start_position + self.content.len() + 2,
                },
            }),
            url: self.content.clone(),
            title: Some(self.content.clone()),
        };
        Node::Link(link)
    }
}

fn extract_links(input: &str) -> Vec<Extracted> {
    let re = Regex::new(r"\[\[([\s\S]*?)\]\]").unwrap();

    input
        .lines()
        .enumerate()
        .flat_map(|(line_number, line)| {
            re.captures_iter(line).filter_map(move |captures| {
                captures.get(1).map(|content| Extracted {
                    content: content.as_str().to_string(),
                    start_position: content.start() + 1,
                    line_number,
                })
            })
        })
        .collect()
}

pub fn parse_wiki_links(node: &mut Node) {
    let mut links = Vec::new();
    if let Some(children) = node.children() {
        for child in children {
            if let Node::Text(t) = child {
                if t.value.contains("[[") && t.value.contains("]]") {
                    let t_position = t.position.as_ref().unwrap().clone();
                    let extracted = extract_links(&t.value);

                    for i in extracted {
                        let link_ast = i.link_node(t_position.start.line);
                        links.push(link_ast);
                    }
                }
            }
        }
    }
    if let Some(children) = node.children_mut() {
        children.append(&mut links);
    }
    // recurse through children
    if let Some(children) = node.children_mut() {
        for child in children {
            parse_wiki_links(child);
        }
    }
}

pub fn find_heading_for_url<'a>(node: &'a Node, link_url: &str) -> Option<&'a Heading> {
    if let Node::Heading(heading) = node {
        if let Some(Node::Text(Text { value, .. })) = heading.children.first() {
            if value == &link_url.replace('#', "") {
                return Some(heading);
            }
        }
    };

    // recurse through children
    if let Some(children) = node.children() {
        for child in children {
            if let Some(n) = find_heading_for_url(child, link_url) {
                return Some(n);
            }
        }
    }
    None
}

pub fn find_link_for_position(node: &Node, line: u32, character: u32) -> Option<&Node> {
    match node {
        Node::Link(Link { position, .. })
        | Node::LinkReference(LinkReference { position, .. })
        | Node::FootnoteReference(FootnoteReference { position, .. }) => {
            if let Some(pos) = position {
                if (line + 1) as usize >= pos.start.line
                    && (line + 1) as usize <= pos.end.line
                    && ((character + 1) as usize) >= pos.start.column
                    && ((character + 1) as usize) <= pos.end.column
                {
                    return Some(node);
                }
            }
        }
        _ => {}
    };

    // recurse through children
    if let Some(children) = node.children() {
        for child in children {
            if let Some(n) = find_link_for_position(child, line, character) {
                return Some(n);
            }
        }
    }
    None
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_make_wiki_links() {
        let md = "Wikilink [[link]] in paragraph";
        let mut ast = markdown::to_mdast(md, &markdown::ParseOptions::gfm()).unwrap();
        // dbg!(&ast);
        parse_wiki_links(&mut ast);
        // dbg!(&ast);
    }

    #[test]
    fn test_make_wiki_links2() {
        let md = "Another line with [[link1]]
Wikilink [[link]] in paragraph ";
        let mut ast = markdown::to_mdast(md, &markdown::ParseOptions::gfm()).unwrap();
        dbg!(&ast);
        parse_wiki_links(&mut ast);
        dbg!(&ast);
    }
}
