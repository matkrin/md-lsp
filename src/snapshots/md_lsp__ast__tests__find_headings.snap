---
source: src/ast.rs
expression: headings
---
[
    Heading {
        children: [
            Text {
                value: "Some Markdown features (github flavored)",
                position: Some(
                    1:3-1:43 (2-42),
                ),
            },
        ],
        position: Some(
            1:1-1:43 (0-42),
        ),
        depth: 1,
    },
    Heading {
        children: [
            Text {
                value: "Heading 1",
                position: Some(
                    31:3-31:12 (857-866),
                ),
            },
        ],
        position: Some(
            31:1-31:12 (855-866),
        ),
        depth: 1,
    },
    Heading {
        children: [
            Text {
                value: "Heading 2",
                position: Some(
                    32:4-32:13 (870-879),
                ),
            },
        ],
        position: Some(
            32:1-32:13 (867-879),
        ),
        depth: 2,
    },
    Heading {
        children: [
            Text {
                value: "Heading 3",
                position: Some(
                    33:5-33:14 (884-893),
                ),
            },
        ],
        position: Some(
            33:1-33:14 (880-893),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Subscript",
                position: Some(
                    43:5-43:14 (993-1002),
                ),
            },
        ],
        position: Some(
            43:1-43:14 (989-1002),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Superscript",
                position: Some(
                    47:5-47:16 (1024-1035),
                ),
            },
        ],
        position: Some(
            47:1-47:16 (1020-1035),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Lists",
                position: Some(
                    51:5-51:10 (1085-1090),
                ),
            },
        ],
        position: Some(
            51:1-51:10 (1081-1090),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Unordered",
                position: Some(
                    53:6-53:15 (1097-1106),
                ),
            },
        ],
        position: Some(
            53:1-53:15 (1092-1106),
        ),
        depth: 4,
    },
    Heading {
        children: [
            Text {
                value: "Ordered",
                position: Some(
                    59:6-59:13 (1132-1139),
                ),
            },
        ],
        position: Some(
            59:1-59:13 (1127-1139),
        ),
        depth: 4,
    },
    Heading {
        children: [
            Text {
                value: "Nested",
                position: Some(
                    64:6-64:12 (1161-1167),
                ),
            },
        ],
        position: Some(
            64:1-64:12 (1156-1167),
        ),
        depth: 4,
    },
    Heading {
        children: [
            Text {
                value: "Task List",
                position: Some(
                    70:5-70:14 (1217-1226),
                ),
            },
        ],
        position: Some(
            70:1-70:14 (1213-1226),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Code",
                position: Some(
                    76:5-76:9 (1269-1273),
                ),
            },
        ],
        position: Some(
            76:1-76:9 (1265-1273),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Inline",
                position: Some(
                    78:6-78:12 (1280-1286),
                ),
            },
        ],
        position: Some(
            78:1-78:12 (1275-1286),
        ),
        depth: 4,
    },
    Heading {
        children: [
            Text {
                value: "Block",
                position: Some(
                    82:6-82:11 (1326-1331),
                ),
            },
        ],
        position: Some(
            82:1-82:11 (1321-1331),
        ),
        depth: 4,
    },
    Heading {
        children: [
            Text {
                value: "Quotes",
                position: Some(
                    90:5-90:11 (1389-1395),
                ),
            },
        ],
        position: Some(
            90:1-90:11 (1385-1395),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Links",
                position: Some(
                    94:5-94:10 (1429-1434),
                ),
            },
        ],
        position: Some(
            94:1-94:10 (1425-1434),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Images",
                position: Some(
                    98:5-98:11 (1512-1518),
                ),
            },
        ],
        position: Some(
            98:1-98:11 (1508-1518),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Footnotes",
                position: Some(
                    103:5-103:14 (1784-1793),
                ),
            },
        ],
        position: Some(
            103:1-103:14 (1780-1793),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Tables",
                position: Some(
                    116:5-116:11 (2010-2016),
                ),
            },
        ],
        position: Some(
            116:1-116:11 (2006-2016),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Colors",
                position: Some(
                    124:5-124:11 (2123-2129),
                ),
            },
        ],
        position: Some(
            124:1-124:11 (2119-2129),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Alerts",
                position: Some(
                    128:5-128:11 (2190-2196),
                ),
            },
        ],
        position: Some(
            128:1-128:11 (2186-2196),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "Comments",
                position: Some(
                    145:5-145:13 (2589-2597),
                ),
            },
        ],
        position: Some(
            145:1-145:13 (2585-2597),
        ),
        depth: 3,
    },
    Heading {
        children: [
            Text {
                value: "LinkRefernce and Definition",
                position: Some(
                    149:5-149:32 (2656-2683),
                ),
            },
        ],
        position: Some(
            149:1-149:32 (2652-2683),
        ),
        depth: 3,
    },
]
